src/Build/Utilities/EngineFileUtilities.cs

```diff
@@ -152,8 +219,193 @@ FileMatcher fileMatcher
                 // the list into a string array.  If the filespec started out
                 // as a relative path, we will get back a bunch of relative paths.
                 // If the filespec started out as an absolute path, we will get
-                // back a bunch of absolute paths.
-                fileList = fileMatcher.GetFiles(directoryUnescaped, filespecUnescaped, excludeSpecsUnescaped);
+                // back a bunch of absolute paths. Also retrieves the search action
+                // and relevant Exclude filespec for drive enumerating wildcard detection.
+                (fileList, action, excludeFileSpec) = fileMatcher.GetFiles(directoryUnescaped, filespecUnescaped, excludeSpecsUnescaped);
+
+                // Determines whether Exclude filespec or passed in file spec should be
+                // used in drive enumeration warning or exception.
+                bool excludeFileSpecIsUnusable = string.IsNullOrWhiteSpace(excludeFileSpec);
+                string fileSpec = excludeFileSpecIsUnusable ? filespecUnescaped : excludeFileSpec;
+
+                switch (action)
```

Comment:
Is there a more optimal way to clean up the logic within these switch statements? I tried to reduce as many lines of code as possible.

Also, could someone double check to make certain that I have the correct conditions in place here? I want to make certain that I'm not causing any false positive warnings or errors (in spite of test coverage). This section might be more clear once I clean up test coverage and add more inline data in `ProjectItem_Tests.cs` and `ProjectItemInstance_Tests.cs`.

Reply:
> Is there a more optimal way to clean up the logic within these switch statements? I tried to reduce as many lines of code as possible.

Just a few random ideas:
1. You could try to flatten the code to have only one switch with pattern matching. Something like: https://dotnetfiddle.net/DUHvNn
2. Or maybe it would be better to refactor the inner switches into their own helper method each.

Reply:
I haven't looked to see if this is viable in this case, but it can sometimes be fairly clean to have two separate switches, one to select relevant arguments and one to select a delegate to apply to them. Then you can just apply it at the end.

Reply:
I went ahead with approach 2 to make it less clunky in the inner switch statements.